File: mellsell-frontend/src/utils/inputSanitizer.js
/**
 * Validação de texto alinhada ao backend (InputSanitizer.java).
 * Bloqueia fragmentos de injeção SQL/XSS em campos de formulário.
 */

export const FORBIDDEN_FRAGMENTS = [
  '--',
  ';',
  '/*',
  '*/',
  '@@',
  'char(',
  'nchar(',
  'varchar(',
  'nvarchar(',
  'xp_',
  '<script',
]

const SQL_KEYWORD = /\b(alter|create|delete|drop|exec|execute|insert|select|union|update)\b/i

const SAFE_NAME = /^[\p{L}0-9 .'-]{2,120}$/u
const SAFE_STORE = /^[\p{L}0-9 .'&-]{0,120}$/u
const SAFE_PRODUCT_NAME = /^[\p{L}0-9 .'&-]{2,120}$/u
const SAFE_ADDRESS_LINE = /^[\p{L}0-9 .,'/ºª°#-]{0,120}$/u
const SAFE_ADDRESS_NUMBER = /^[\p{L}0-9 ./-]{1,20}$/u
const SAFE_SHIPPING = /^[\p{L}0-9 .,'/ºª°#()\n—-]{10,500}$/u
const SAFE_DESCRIPTION = /^[\p{L}0-9 .,!?'\n\r-]{0,2000}$/u
const SAFE_REVIEW = /^[\p{L}0-9 .,!?'\n\r-]{1,1000}$/u

export function hasDangerousText(value) {
  const text = String(value ?? '')
  if (!text) return false
  const lower = text.toLowerCase()
  if (FORBIDDEN_FRAGMENTS.some((f) => lower.includes(f))) return true
  if (SQL_KEYWORD.test(text)) return true
  if (/[<>]/.test(text)) return true
  return false
}

export function getDangerousTextError(fieldLabel = 'Campo') {
  return `${fieldLabel} contém texto não permitido.`
}

export function getSafeNameError(value) {
  const trimmed = String(value ?? '').trim().replace(/\s+/g, ' ')
  if (!trimmed) return 'Nome completo é obrigatório.'
  if (hasDangerousText(trimmed)) return getDangerousTextError('Nome')
  if (!SAFE_NAME.test(trimmed)) {
    return 'Nome contém caracteres não permitidos. Use apenas letras, números, espaços e hífen.'
  }
  if (!trimmed.includes(' ')) {
    return 'Informe o nome completo (nome e sobrenome, separados por espaço).'
  }
  const parts = trimmed.split(' ')
  if (parts.length < 2) return 'Informe pelo menos nome e sobrenome.'
  if (parts.some((p) => p.length < 2)) {
    return 'Cada parte do nome deve ter pelo menos 2 caracteres.'
  }
  return null
}

export function getSafeStoreNameError(value) {
  const trimmed = String(value ?? '').trim()
  if (!trimmed) return null
  if (hasDangerousText(trimmed)) return getDangerousTextError('Nome da loja')
  if (!SAFE_STORE.test(trimmed)) return 'Nome da loja contém caracteres não permitidos.'
  return null
}

/** Nome da loja obrigatório no cadastro de apicultor/fornecedor. */
export function getVendorStoreNameError(value) {
  const trimmed = String(value ?? '').trim()
  if (!trimmed) return 'Informe o nome da sua loja ou apiário.'
  if (trimmed.length < 3) return 'Nome da loja deve ter pelo menos 3 caracteres.'
  return getSafeStoreNameError(trimmed)
}

/** Descrição da loja no cadastro de apicultor/fornecedor. */
export function getVendorDescriptionError(value) {
  const trimmed = String(value ?? '').trim()
  if (!trimmed) return 'Descreva sua loja ou apiário.'
  if (trimmed.length < 20) return 'Descrição deve ter pelo menos 20 caracteres.'
  return getSafeProductDescriptionError(trimmed)
}

/** Cidade no cadastro de apicultor/fornecedor. */
export function getVendorCityError(value) {
  const trimmed = String(value ?? '').trim().replace(/\s+/g, ' ')
  if (!trimmed) return 'Cidade é obrigatória.'
  if (trimmed.length < 2) return 'Cidade deve ter pelo menos 2 caracteres.'
  return getSafeAddressLineError(trimmed, 'Cidade', false)
}

/** UF no cadastro de apicultor/fornecedor. */
export function getVendorStateError(value) {
  const state = String(value ?? '').trim().toUpperCase()
  if (!state) return 'UF é obrigatória.'
  if (hasDangerousText(state)) return getDangerousTextError('UF')
  if (!/^[A-Z]{2}$/.test(state)) return 'UF deve ter 2 letras (ex.: SP).'
  return null
}

export function getSafeProductNameError(value) {
  const trimmed = String(value ?? '').trim().replace(/\s+/g, ' ')
  if (!trimmed) return 'Nome do produto é obrigatório.'
  if (hasDangerousText(trimmed)) return getDangerousTextError('Nome do produto')
  if (!SAFE_PRODUCT_NAME.test(trimmed)) {
    return 'Nome do produto contém caracteres não permitidos.'
  }
  return null
}

export function getSafeProductDescriptionError(value) {
  const trimmed = String(value ?? '').trim()
  if (!trimmed) return null
  if (hasDangerousText(trimmed)) return getDangerousTextError('Descrição')
  if (!SAFE_DESCRIPTION.test(trimmed)) return 'Descrição contém caracteres não permitidos.'
  return null
}

export function getSafeReviewCommentError(value) {
  const trimmed = String(value ?? '').trim()
  if (!trimmed) return null
  if (hasDangerousText(trimmed)) return getDangerousTextError('Comentário')
  if (!SAFE_REVIEW.test(trimmed)) return 'Comentário contém caracteres não permitidos.'
  return null
}

export function getSafeAddressLineError(value, label, required = false) {
  const trimmed = String(value ?? '').trim().replace(/\s+/g, ' ')
  if (!trimmed) return required ? `${label} é obrigatório.` : null
  if (hasDangerousText(trimmed)) return getDangerousTextError(label)
  if (!SAFE_ADDRESS_LINE.test(trimmed)) return `${label} contém caracteres não permitidos.`
  return null
}

export function getSafeAddressNumberError(value) {
  const trimmed = String(value ?? '').trim()
  if (!trimmed) return 'Número é obrigatório.'
  if (hasDangerousText(trimmed)) return getDangerousTextError('Número')
  if (!SAFE_ADDRESS_NUMBER.test(trimmed)) return 'Número contém caracteres não permitidos.'
  return null
}

export function getSafeShippingAddressError(value) {
  const trimmed = String(value ?? '').trim().replace(/[ \t]+/g, ' ')
  if (!trimmed) return 'Endereço de entrega é obrigatório.'
  if (hasDangerousText(trimmed)) return getDangerousTextError('Endereço')
  if (!SAFE_SHIPPING.test(trimmed)) return 'Endereço contém caracteres não permitidos.'
  return null
}

/** Valida campos do formulário de CEP antes do checkout. */
export function getAddressFieldsValidationError(fields) {
  const cepDigits = String(fields?.cep ?? '').replace(/\D/g, '')
  if (cepDigits.length !== 8) return 'Informe um CEP com 8 dígitos.'
  if (hasDangerousText(cepDigits)) return getDangerousTextError('CEP')

  const streetErr = getSafeAddressLineError(fields?.street, 'Logradouro', true)
  if (streetErr) return streetErr

  const numberErr = getSafeAddressNumberError(fields?.number)
  if (numberErr) return numberErr

  const complementErr = getSafeAddressLineError(fields?.complement, 'Complemento', false)
  if (complementErr) return complementErr

  const neighborhoodErr = getSafeAddressLineError(fields?.neighborhood, 'Bairro', true)
  if (neighborhoodErr) return neighborhoodErr

  const cityErr = getSafeAddressLineError(fields?.city, 'Cidade', true)
  if (cityErr) return cityErr

  const state = String(fields?.state ?? '').trim()
  if (!state) return 'UF é obrigatória.'
  if (hasDangerousText(state)) return getDangerousTextError('UF')
  if (!/^[A-Za-z]{2}$/.test(state)) return 'UF inválida.'

  return null
}

/** Remove caracteres de marcação HTML comuns em inputs de texto. */
export function stripMarkupChars(value) {
  return String(value ?? '').replace(/[<>]/g, '')
}

export function validateProductForm({ name, description }) {
  return getSafeProductNameError(name) || getSafeProductDescriptionError(description)
}

File: mellsell-backend/src/main/java/com/mellsell/common/util/InputSanitizer.java
package com.mellsell.common.util;

import java.util.regex.Pattern;

/**
 * Validação de texto de entrada — complementa JPA (consultas parametrizadas) bloqueando padrões de injeção.
 */
public final class InputSanitizer {

    private static final Pattern SAFE_NAME = Pattern.compile("^[\\p{L}0-9 .'\\-]{2,120}$");
    private static final Pattern SAFE_STORE = Pattern.compile("^[\\p{L}0-9 .'\\-&]{0,120}$");
    private static final Pattern SAFE_PRODUCT_NAME = Pattern.compile("^[\\p{L}0-9 .'\\-&]{2,120}$");
    private static final Pattern SAFE_ADDRESS_LINE = Pattern.compile("^[\\p{L}0-9 .,'\\-/ºª°#]{0,120}$");
    private static final Pattern SAFE_ADDRESS_NUMBER = Pattern.compile("^[\\p{L}0-9 ./\\-]{1,20}$");
    private static final Pattern SAFE_SHIPPING = Pattern.compile("^[\\p{L}0-9 .,'\\-/ºª°#()\\n—]{10,500}$");
    private static final Pattern CEP_DIGITS = Pattern.compile("^\\d{8}$");
    private static final Pattern SAFE_DESCRIPTION = Pattern.compile("^[\\p{L}0-9 .,!?'\\-\\n\\r]{0,2000}$");
    private static final Pattern SAFE_REVIEW = Pattern.compile("^[\\p{L}0-9 .,!?'\\-\\n\\r]{1,1000}$");
    /** Palavras SQL inteiras — evita falsos positivos em nomes como "Valter", "Walter". */
    private static final Pattern SQL_KEYWORD = Pattern.compile(
            "\\b(alter|create|delete|drop|exec|execute|insert|select|union|update)\\b",
            Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE
    );
    private static final String[] FORBIDDEN_FRAGMENTS = {
            "--", ";", "/*", "*/", "@@", "char(", "nchar(", "varchar(", "nvarchar(", "xp_", "<script"
    };

    private InputSanitizer() {}

    public static String requireSafeName(String name) {
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("Nome completo é obrigatório.");
        }
        String trimmed = name.trim().replaceAll("\\s+", " ");
        rejectDangerousFragments(trimmed, "Nome");
        if (!SAFE_NAME.matcher(trimmed).matches()) {
            throw new IllegalArgumentException("Nome contém caracteres não permitidos. Use apenas letras, números, espaços e hífen.");
        }
        if (!trimmed.contains(" ")) {
            throw new IllegalArgumentException("Informe o nome completo (nome e sobrenome, separados por espaço).");
        }
        String[] parts = trimmed.split(" ");
        if (parts.length < 2) {
            throw new IllegalArgumentException("Informe pelo menos nome e sobrenome.");
        }
        for (String part : parts) {
            if (part.length() < 2) {
                throw new IllegalArgumentException("Cada parte do nome deve ter pelo menos 2 caracteres.");
            }
        }
        return trimmed;
    }

    public static String safeStoreName(String storeName) {
        if (storeName == null || storeName.isBlank()) {
            return null;
        }
        String trimmed = storeName.trim();
        rejectDangerousFragments(trimmed, "Nome da loja");
        if (!SAFE_STORE.matcher(trimmed).matches()) {
            throw new IllegalArgumentException("Nome da loja contém caracteres não permitidos.");
        }
        return trimmed;
    }

    public static String normalizeEmail(String email) {
        if (email == null || email.isBlank()) {
            throw new IllegalArgumentException("E-mail é obrigatório.");
        }
        String trimmed = email.trim().toLowerCase();
        rejectDangerousFragments(trimmed, "E-mail");
        return trimmed;
    }

    public static String requireSafeCep(String rawCep) {
        if (rawCep == null) {
            throw new IllegalArgumentException("CEP é obrigatório.");
        }
        String digits = rawCep.replaceAll("\\D", "");
        rejectDangerousFragments(digits, "CEP");
        if (!CEP_DIGITS.matcher(digits).matches()) {
            throw new IllegalArgumentException("CEP deve conter 8 dígitos.");
        }
        return digits;
    }

    public static String requireSafeAddressLine(String line, String fieldLabel, boolean required) {
        if (line == null || line.isBlank()) {
            if (required) {
                throw new IllegalArgumentException(fieldLabel + " é obrigatório.");
            }
            return "";
        }
        String trimmed = line.trim().replaceAll("\\s+", " ");
        rejectDangerousFragments(trimmed, fieldLabel);
        if (!SAFE_ADDRESS_LINE.matcher(trimmed).matches()) {
            throw new IllegalArgumentException(fieldLabel + " contém caracteres não permitidos.");
        }
        return trimmed;
    }

    public static String requireSafeAddressNumber(String number) {
        if (number == null || number.isBlank()) {
            throw new IllegalArgumentException("Número é obrigatório.");
        }
        String trimmed = number.trim();
        rejectDangerousFragments(trimmed, "Número");
        if (!SAFE_ADDRESS_NUMBER.matcher(trimmed).matches()) {
            throw new IllegalArgumentException("Número contém caracteres não permitidos.");
        }
        return trimmed;
    }

    public static String requireSafeShippingAddress(String address) {
        if (address == null || address.isBlank()) {
            throw new IllegalArgumentException("Endereço de entrega é obrigatório.");
        }
        String trimmed = address.trim().replaceAll("[ \\t]+", " ").replaceAll("\\n{3,}", "\n\n");
        rejectDangerousFragments(trimmed, "Endereço");
        if (!SAFE_SHIPPING.matcher(trimmed).matches()) {
            throw new IllegalArgumentException("Endereço contém caracteres não permitidos.");
        }
        return trimmed;
    }

    public static String requireSafeProductName(String name) {
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("Nome do produto é obrigatório.");
        }
        String trimmed = name.trim().replaceAll("\\s+", " ");
        rejectDangerousFragments(trimmed, "Nome do produto");
        if (!SAFE_PRODUCT_NAME.matcher(trimmed).matches()) {
            throw new IllegalArgumentException("Nome do produto contém caracteres não permitidos.");
        }
        return trimmed;
    }

    public static String safeProductDescription(String description) {
        if (description == null || description.isBlank()) {
            return "";
        }
        String trimmed = description.trim();
        rejectDangerousFragments(trimmed, "Descrição");
        if (!SAFE_DESCRIPTION.matcher(trimmed).matches()) {
            throw new IllegalArgumentException("Descrição contém caracteres não permitidos.");
        }
        return trimmed;
    }

    public static String requireSafeBrazilianState(String state) {
        if (state == null || state.isBlank()) {
            throw new IllegalArgumentException("UF é obrigatória.");
        }
        String trimmed = state.trim().toUpperCase();
        rejectDangerousFragments(trimmed, "UF");
        if (!Pattern.compile("^[A-Z]{2}$").matcher(trimmed).matches()) {
            throw new IllegalArgumentException("UF inválida.");
        }
        return trimmed;
    }

    public static String requireSafeStoreName(String storeName) {
        if (storeName == null || storeName.isBlank()) {
            throw new IllegalArgumentException("Informe o nome da sua loja ou apiário");
        }
        String trimmed = storeName.trim();
        if (trimmed.length() < 3) {
            throw new IllegalArgumentException("Nome da loja deve ter pelo menos 3 caracteres");
        }
        rejectDangerousFragments(trimmed, "Nome da loja");
        if (!SAFE_STORE.matcher(trimmed).matches()) {
            throw new IllegalArgumentException("Nome da loja contém caracteres não permitidos.");
        }
        return trimmed;
    }

    public static String safeReviewComment(String comment) {
        if (comment == null || comment.isBlank()) {
            return null;
        }
        String trimmed = comment.trim();
        rejectDangerousFragments(trimmed, "Comentário");
        if (!SAFE_REVIEW.matcher(trimmed).matches()) {
            throw new IllegalArgumentException("Comentário contém caracteres não permitidos.");
        }
        return trimmed;
    }

    public static String requireSafeSupplierDescription(String description) {
        if (description == null || description.isBlank()) {
            throw new IllegalArgumentException("Descreva sua loja ou apiário.");
        }
        String trimmed = description.trim();
        if (trimmed.length() < 20) {
            throw new IllegalArgumentException("Descrição deve ter pelo menos 20 caracteres.");
        }
        rejectDangerousFragments(trimmed, "Descrição da loja");
        if (!SAFE_DESCRIPTION.matcher(trimmed).matches()) {
            throw new IllegalArgumentException("Descrição da loja contém caracteres não permitidos.");
        }
        return trimmed;
    }

    public static String requireSafeSupplierCity(String city) {
        if (city == null || city.isBlank()) {
            throw new IllegalArgumentException("Cidade é obrigatória.");
        }
        String trimmed = city.trim().replaceAll("\\s+", " ");
        if (trimmed.length() < 2) {
            throw new IllegalArgumentException("Cidade deve ter pelo menos 2 caracteres.");
        }
        rejectDangerousFragments(trimmed, "Cidade");
        if (!SAFE_ADDRESS_LINE.matcher(trimmed).matches()) {
            throw new IllegalArgumentException("Cidade contém caracteres não permitidos.");
        }
        return trimmed;
    }

    public static String requireSafeSupplierState(String state) {
        if (state == null || state.isBlank()) {
            throw new IllegalArgumentException("UF é obrigatória.");
        }
        String trimmed = state.trim().toUpperCase();
        rejectDangerousFragments(trimmed, "UF");
        if (!trimmed.matches("^[A-Z]{2}$")) {
            throw new IllegalArgumentException("UF deve ter 2 letras (ex.: SP).");
        }
        return trimmed;
    }

    private static void rejectDangerousFragments(String value, String fieldLabel) {
        String lower = value.toLowerCase();
        for (String fragment : FORBIDDEN_FRAGMENTS) {
            if (lower.contains(fragment)) {
                throw new IllegalArgumentException(fieldLabel + " contém texto não permitido.");
            }
        }
        if (SQL_KEYWORD.matcher(value).find()) {
            throw new IllegalArgumentException(fieldLabel + " contém texto não permitido.");
        }
    }
}